huaweicloud/services/eip/data_source_huaweicloud_vpc_internet_gateways.go

```diff
@@ -0,0 +1,159 @@
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"github.com/chnsz/golangsdk"
+
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/config"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/utils"
+)
+
+// @API EIP GET /v3/{project_id}/geip/vpc-igws
+func DataSourceVPCInternetGateways() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: dataSourceVPCInternetGatewaysRead,
+
+		Schema: map[string]*schema.Schema{
+			"region": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"igw_name": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"igw_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"vpc_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"vpc_igws": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"name": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"vpc_id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"subnet_id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"enable_ipv6": {
+							Type:     schema.TypeBool,
+							Computed: true,
+						},
+						"created_at": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"updated_at": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func dataSourceVPCInternetGatewaysRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	region := cfg.GetRegion(d)
+	client, err := cfg.NetworkingV3Client(region)
+	if err != nil {
+		return diag.Errorf("error creating VPC V3 client: %s", err)
+	}
+
+	getInternetGatewaysHttpUrl := "v3/{project_id}/geip/vpc-igws"
+	getInternetGatewaysPath := client.Endpoint + getInternetGatewaysHttpUrl
+	getInternetGatewaysPath = strings.ReplaceAll(getInternetGatewaysPath, "{project_id}", client.ProjectID)
+	getInternetGatewaysOpt := golangsdk.RequestOpts{
+		KeepResponseBody: true,
+	}
+
+	getInternetGatewaysPath += "?limit=10"
+	if igwName, ok := d.GetOk("igw_name"); ok {
+		getInternetGatewaysPath += fmt.Sprintf("&name=%s", igwName)
+	}
+	if igwID, ok := d.GetOk("igw_id"); ok {
+		getInternetGatewaysPath += fmt.Sprintf("&id=%s", igwID)
+	}
+	if vpcID, ok := d.GetOk("vpc_id"); ok {
+		getInternetGatewaysPath += fmt.Sprintf("&vpc_id=%s", vpcID)
+	}
+	currentTotal := 0
+
+	// Although the offset is not in HuaweiCloud API docs, but it works actually.
+	getInternetGatewaysPath += fmt.Sprintf("&offset=%v", currentTotal)
+
+	results := make([]map[string]interface{}, 0)
+	for {
+		getInternetGatewaysResp, err := client.Request("GET", getInternetGatewaysPath, &getInternetGatewaysOpt)
```

Comment:
it's better to generate the API path in loop.
```suggestion

	currentTotal := 0
	results := make([]map[string]interface{}, 0)
	for {
		// Although the offset is not in HuaweiCloud API docs, but it works actually.
		currentPath := getInternetGatewaysPath + fmt.Sprintf("&offset=%d", currentTotal)
		getInternetGatewaysResp, err := client.Request("GET", currentPath, &getInternetGatewaysOpt)
```